pkg/cluster/healthcheck/healthcheck.go

```diff
@@ -0,0 +1,282 @@
+		healthyThreshold:   healthyThreshold,
+		unhealthyThreshold: unhealthyThreshold,
+		initialDelay:       initialDelay,
+		checkers:           make(map[string]*EndpointChecker),
+	}
+
+	return hc
+}
+
+func (hc *HealthChecker) Start() {
+	// each endpoint
+	for _, h := range hc.cluster.Endpoints {
+		hc.startCheck(h)
+	}
+}
+
+func (hc *HealthChecker) Stop() {
+	for _, h := range hc.cluster.Endpoints {
+		hc.stopCheck(h)
+	}
+}
+
+func (hc *HealthChecker) StopOne(endpoint *model.Endpoint) {
+	hc.stopCheck(endpoint)
+}
+
+func (hc *HealthChecker) StartOne(endpoint *model.Endpoint) {
+	hc.startCheck(endpoint)
+}
+
+func (hc *HealthChecker) startCheck(endpoint *model.Endpoint) {
+	addr := endpoint.Address.GetAddress()
+	if _, ok := hc.checkers[addr]; !ok {
+		c := newChecker(endpoint, hc)
+		hc.checkers[addr] = c
+		c.Start()
+		logger.Infof("[health check] create a health check session for %s", addr)
+	}
+}
+
+func (hc *HealthChecker) stopCheck(endpoint *model.Endpoint) {
+	addr := endpoint.Address.GetAddress()
+	if c, ok := hc.checkers[addr]; ok {
+		c.Stop()
+		delete(hc.checkers, addr)
+		logger.Infof("[health check] create a health check session for %s", addr)
+	}
+}
+
+func newChecker(endpoint *model.Endpoint, hc *HealthChecker) *EndpointChecker {
+	c := &EndpointChecker{
+		tcpChecker:    newTcpChecker(endpoint, hc.timeout),
+		endpoint:      endpoint,
+		HealthChecker: hc,
+		resp:          make(chan checkResponse),
+		timeout:       make(chan bool),
+		stop:          make(chan struct{}),
+	}
+	return c
+}
+
+func newTcpChecker(endpoint *model.Endpoint, timeout time.Duration) *TCPChecker {
+	return &TCPChecker{
+		addr:    endpoint.Address.GetAddress(),
+		timeout: timeout,
+	}
+}
+
+func (hc *HealthChecker) getCheckInterval() time.Duration {
+	return hc.intervalBase
+}
+
+func (c *EndpointChecker) Start() {
+	defer func() {
+		if r := recover(); r != nil {
+			logger.Warnf("[health check] node checker panic %v\n%s", r, string(debug.Stack()))
+		}
+		c.checkTimer.Stop()
+		c.checkTimeout.Stop()
+	}()
+	c.checkTimer = gxtime.AfterFunc(c.HealthChecker.initialDelay, c.OnCheck)
+	for {
+		select {
+		case <-c.stop:
+			return
+		default:
+			// prepare a check
+			currentID := atomic.AddUint64(&c.checkID, 1)
+			select {
+			case <-c.stop:
+				return
+			case resp := <-c.resp:
+				if resp.ID == currentID {
+					if c.checkTimeout != nil {
+						c.checkTimeout.Stop()
+						c.checkTimeout = nil
+					}
+					if resp.Healthy {
+						c.HandleSuccess()
+					} else {
+						c.HandleFailure(false)
+					}
+					c.checkTimer = gxtime.AfterFunc(c.HealthChecker.getCheckInterval(), c.OnCheck)
+				}
+			case <-c.timeout:
+				if c.checkTimer != nil {
+					c.checkTimer.Stop()
+				}
+				c.HandleFailure(true)
+				c.checkTimer = gxtime.AfterFunc(c.HealthChecker.getCheckInterval(), c.OnCheck)
+				logger.Infof("[health check] receive a timeout response at id: %d", currentID)
+
+			}
+		}
+	}
+}
+
+func (c *EndpointChecker) Stop() {
+	close(c.stop)
```

Comment:
close a closed chan cause a panic， considerate recover?

Reply:
use once, will be deleted from checker list after Stop